File: Association Slope/Hash/Q2_Copy List with Random Pointer/Q2_CopyListwithRandomPointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x, next=None, random=None):
        self.val = int(x)   # store the value of the node
        self.next = next    # pointer to the next node
        self.random = random # pointer to a random node
"""

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        # If the original list is empty, return None
        if head is None:
            return None  
        
        # Dictionary to store mapping from original node -> copied node
        node_map = {}
        
        # First pass: create a copy of each node and store in the dictionary
        current = head
        while current:
            node_map[current] = Node(current.val)  # create a new node with the same value
            current = current.next                 # move to the next node in original list
        
        # Second pass: assign next and random pointers for copied nodes
        current = head
        while current:
            copy = node_map[current]                   # get the copied node corresponding to current
            copy.next = node_map.get(current.next)     # set the next pointer of copied node
            copy.random = node_map.get(current.random) # set the random pointer of copied node
            current = current.next                     # move to the next node in original list
        
        # Return the head of the copied linked list
        return node_map[head]

File: Association Slope/Hash/Q2_Copy List with Random Pointer/Q2_CopyListwithRandomPointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

import java.util.HashMap;

import org.w3c.dom.Node;

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null; // handle empty list
        
        // Map to store original node -> copied node
        HashMap<Node, Node> map = new HashMap<>();
        
        // First pass: create all nodes and store in map
        Node current = head;
        while (current != null) {
            map.put(current, new Node(current.val));
            current = current.next;
        }
        
        // Second pass: set next and random pointers
        current = head;
        while (current != null) {
            Node copy = map.get(current);              // get the copied node
            copy.next = map.get(current.next);        // set next pointer
            copy.random = map.get(current.random);    // set random pointer
            current = current.next;
        }
        
        // Return the head of the copied list
        return map.get(head);
    }
}
